src/ynEIGEN/TokenStakingNode.sol

```diff
@@ -108,6 +129,122 @@ contract TokenStakingNode is
+
+        _fullWithdrawalRoots = tokenStakingNodesManager.delegationManager().queueWithdrawals(_params);
+
+        emit QueuedWithdrawals(_strategy, _shares, _fullWithdrawalRoots);
+    }
+
+    function completeQueuedWithdrawals(
+        uint256 _nonce,
+        uint32 _startBlock,
+        uint256 _shares,
+        IStrategy _strategy,
+        uint256[] memory _middlewareTimesIndexes
+    ) public onlyTokenStakingNodesWithdrawer {
+
+        IDelegationManager _delegationManager = tokenStakingNodesManager.delegationManager();
+
+        IDelegationManager.Withdrawal[] memory _withdrawals = new IDelegationManager.Withdrawal[](1);
+        {
+            IStrategy[] memory _strategiesArray = new IStrategy[](1);
+            _strategiesArray[0] = _strategy;
+            uint256[] memory _sharesArray = new uint256[](1);
+            _sharesArray[0] = _shares;
+            _withdrawals[0] = IDelegationManager.Withdrawal({
+                staker: address(this),
+                delegatedTo: _delegationManager.delegatedTo(address(this)),
+                withdrawer: address(this),
+                nonce: _nonce,
+                startBlock: _startBlock,
+                strategies: _strategiesArray,
+                shares: _sharesArray
+            });
+        }
+
+        uint256 _expectedAmountOut = _strategy.sharesToUnderlyingView(_shares);
+
+        IERC20 _token = _strategy.underlyingToken();
+        uint256 _balanceBefore = _token.balanceOf(address(this));
+
+        {
+            bool[] memory _receiveAsTokens = new bool[](1);
+            _receiveAsTokens[0] = true;
+            IERC20[][] memory _tokens = new IERC20[][](1);
+            _tokens[0] = new IERC20[](1);
+            _tokens[0][0] = _token;
+
+            _delegationManager.completeQueuedWithdrawals(
+                _withdrawals,
+                _tokens,
+                _middlewareTimesIndexes,
+                _receiveAsTokens
+            );
+        }
+
+        uint256 _actualAmountOut = _token.balanceOf(address(this)) - _balanceBefore;
+        uint256 _delta = _actualAmountOut > _expectedAmountOut ?
+            _actualAmountOut - _expectedAmountOut :
+            _expectedAmountOut - _actualAmountOut;
+        if (_delta > 2) revert WithdrawalAmountMismatch(); // @todo - might be a footgun
+
+        (_actualAmountOut, _token) = _wrapIfNeeded(_actualAmountOut, _token);
+
+        queuedShares[_strategy] -= _shares;
+        withdrawn[_token] += _actualAmountOut;
+
+        emit CompletedQueuedWithdrawals(_shares, _actualAmountOut, address(_strategy));
+    }
+
+    function deallocateTokens(IERC20 _token, uint256 _amount) external onlyTokenStakingNodesManager {
+
+        withdrawn[_token] -= _amount;
+        _token.forceApprove(address(tokenStakingNodesManager), _amount);
+
+        emit DeallocatedTokens(_amount, _token);
+    }
+
+    function _wrapIfNeeded(uint256 _amount, IERC20 _token) internal returns (uint256, IERC20) {
```

Comment:
we should have this asset specific logic not contained here

currently all those asset specific things are in EigenStrategyManager

We should consider refactoring things so it's all contained in 1 contract